sql/wsrep_async_monitor.cc

```diff
@@ -0,0 +1,94 @@
+/* Copyright (c) 2024 Percona LLC and/or its affiliates. All rights reserved.
+
+   This program is free software; you can redistribute it and/or
+   modify it under the terms of the GNU General Public License
+   as published by the Free Software Foundation; version 2 of
+   the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */
+
+#ifdef WITH_WSREP
+#include "sql/wsrep_async_monitor.h"
+#include <algorithm>
+#include <cassert>
+
+// Method for main thread to add scheduled seqnos
+void Wsrep_async_monitor::schedule(seqno_t seqno) {
+  std::unique_lock<std::mutex> lock(m_mutex);
+  scheduled_seqnos.push(seqno);
+}
+
+// Method for both DDL and DML to enter the monitor
+void Wsrep_async_monitor::enter(seqno_t seqno) {
+  std::unique_lock<std::mutex> lock(m_mutex);
+
+  // Wait until this transaction is at the head of the scheduled queue
+  m_cond.wait(lock, [this, seqno] {
+    // Remove skipped transactions
```

Comment:
When this can happen? Could you add a code comment? 
I suspect this is the case when A, B, C have been scheduled, then B skipped, but skip() removes only from the front of scheduled_seqno queue

Reply:
Yes. Imagine a scenario where scheduled seqnos is 1,2,3 and threads enter in an out of order manner.

- 3 enters the monitor first, since it is not in the front of the queue, it goes to cond_wait
- 2 enters the monitor, adds 2 to `skipped_seqnos` 
- 1 enters the monitor, since it is in the front, it acquires the monitor, does its job and leaves the monitor by removing itself from the `scheduled_seqnos` and signals 3.
- 3 wakes up, but it will see that 2 in the front of the queue. To proceed, it needs to remove all the skipped seqnos in the `scheduled_seqnos` queue.

Reply:
Could you add a comment to the code explaining the purpose of this part?

Reply:
Done